### PR TITLE
Fix SPECIAL-BINDINGS-LET*

### PR DESCRIPTION
**Pull request by appleby**
Fix a bug in `SPECIAL-BINDINGS-LET*` that was causing quilc to crash if the user specified either the `--gate-whitelist` or `--gate-blacklist` options in combination with `-S` or `-P`.

Also:

- use `SPECIAL-BINDINGS-LET*` in `TEST-QUIL-TO-NATIVE-QUIL-ENDPOINT-OVERRIDES-SERVER` rather than `SETF`.
- Remove last traces of hunchentoot from Makefile, build-app.lisp, and quilc.asd.

Copy/paste of commit message:

**Fix a bug in `SPECIAL-BINDINGS-LET*`**

`SPECIAL-BINDINGS-LET*` needs to expand to a form that quotes the value being added to `BT:*DEFAULT-SPECIAL-BINDINGS*`; otherwise, binding non-self-evaluating forms will cause an error once a new thread is created.

In other words, we want to expand to something like:

```
  CL-USER> (macroexpand '(quilc::special-bindings-let* ((*foo* 'foo))))
  (LET* ((*FOO* 'FOO)
         (BORDEAUX-THREADS:*DEFAULT-SPECIAL-BINDINGS*
          (LIST* `(,'*FOO* QUOTE ,*FOO*)
                 BORDEAUX-THREADS:*DEFAULT-SPECIAL-BINDINGS*))))
```

Whereas, previously, we expanded to:

```
  CL-USER> (macroexpand '(quilc::special-bindings-let* ((*foo* 'foo))))
  (LET* ((*FOO* 'FOO)
         (BORDEAUX-THREADS:*DEFAULT-SPECIAL-BINDINGS*
          (LIST* (CONS '*FOO* *FOO*)
                 BORDEAUX-THREADS:*DEFAULT-SPECIAL-BINDINGS*))))
```

### CODE REVIEW

**Comment by appleby** [comment]
Hilariously long/weird backtrace when specifying both `-S` and `--gate-whitelist`, prior to this fix:

```
$ ./quilc --gate-whitelist "RX,RZ" -S
+-----------------+
|  W E L C O M E  |
|   T O   T H E   |
|  R I G E T T I  |
|     Q U I L     |
| C O M P I L E R |
+-----------------+
Copyright (c) 2016-2019 Rigetti Computing.

<134>1 2019-08-19T00:41:36Z BK-LOSX-24JLVCJ.local ./quilc 37115 LOG0001 - Launching quilc.
<134>1 2019-08-19T00:41:36Z BK-LOSX-24JLVCJ.local ./quilc 37115 - - Spawning server at (tcp://*:5555) .

; in: "RX" "RZ"
; in: "RX" "RZ"
; in: "RX" "RZ"
;     ("RX" "RZ")
;     ("RX" "RZ")
;     ("RX" "RZ")
; ;
; ;
; caught ERROR:
"RZ"
; caught ERROR:
; caught ERROR:
; caught ERROR:
;     ("RX" "RZ")
;
;
;   illegal function call
;   illegal function call
; caught ERROR:
;   illegal function call
; in: "RX" "RZ"
;     ("RX" "RZ")
;
; caught ERROR:
;   illegal function call
; in: "RX" "RZ"
;
;     ("RX" "RZ");
;
;
;     ("RX" "RZ");

;
;     ("RX" "RZ");

;
; compilation unit finished
; compilation unit finished
;
; compilation unit finished
; caught ERROR:
;   caught 1 ERROR condition;   caught 1 ERROR condition
;   illegal function call
;   illegal function call
;   caught 1 ERROR condition
; compilation unit finished
;   caught 1 ERROR condition
;
; compilation unit finished
;   caught 1 ERROR condition
Unhandled SB-INT:COMPILED-PROGRAM-ERROR in thread #<SB-THREAD:THREAD "RPC-server-thread-1" RUNNING
Unhandled SB-INT:COMPILED-PROGRAM-ERROR in thread #<SB-THREAD:THREAD "RPC-server-thread-1" RUNNING
Unhandled SB-INT:COMPILED-PROGRAM-ERROR in thread #<SB-THREAD:THREAD "RPC-server-thread-2" RUNNING
                                                     {1004738E93}>:
                                                     {10047392D3}>:
Unhandled SB-INT:COMPILED-PROGRAM-ERROR in thread #<SB-THREAD:THREAD "RPC-server-thread-3" RUNNING

  Execution of a form compiled with errors.  Execution of

Form:
cution of a form compiled with errors.  Execution of
 form compiled with errors.
Form:
cution of a form compiled with errors.  Execution of
 form compiled with errors.
                                                     {1004739713}>:
Form:
         Execution of  Execution of  Execution of a form compiled with errors.
  Execution of a form compiled with errors.
  Execution of a form compiled with errors.
Form:
ution of a form compiled with errors.
  Execution of a form compiled with errors.
Compile-time error:
Form:
Form:
Form:
ution of a form compiled with errors.
  Execution of a form compiled with errors.
Compile-time error:
Compile-time error:
    ("RX" "RZ")


Compile-time error:
 function call


Compile-time error:
Compile-t-time error:
 function call

o  illegal function r  illegal function call  illegal function call

;
;
Backtrace for: e for: Backtrace for: #<#<#<SB-THREADSBSB-THRD:DTHRTHREA"at oP unit fiC-s"R"t
Backtrace for: e for: Backtrace for: #<#<#<SB-THREADSBSB-THRD:DTHRTHREA"at oP unit fiC-s"R"RC-reevrvhea-te2"d a"- 1"0" RUNNING {UNNING100RUN c20043 11004t
Backtrace for: e for: Backtrace for: #<#<#<SB-THREADSBSB-THRD:DTHRTHREA"at oP unit fiC-s"R"RC-reevrvhea-te2"d a"- 1"0" RUNNING {UNNING100RUN c20043 110049R 0o7d8493
t
Backtrace for: e for: Backtrace for: #<#<#<SB-THREADSBSB-THRD:DTHRTHREA"at oP unit fiC-s"R"RC-reevrvhea-te2"d a"- 1"0" RUNNING {UNNING100RUN c20043 110049R 0o7d8493
>
t
Backtrace for: e for: Backtrace for: #<#<#<SB-THREADSBSB-THRD:DTHRTHREA"at oP unit fiC-s"R"RC-reevrvhea-te2"d a"- 1"0" RUNNING {UNNING100RUN c20043 110049R 0o7d8493
>
53}>
t
Backtrace for: e for: Backtrace for: #<#<#<SB-THREADSBSB-THRD:DTHRTHREA"at oP unit fiC-s"R"RC-ree> :QUIT T)
d a"- 1"0" RUNNING {UNNING100RUN c20043 110049R 0o7d8493
>
50: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:COMPILED-PROGRAM-ERROR {10047B9123}> #<unused argument> :QUIT T)
Unhandled SB-INT:COMPILED-PROGRAM-ERROR in thread #<SB-THREAD:THREAD "RPC-server-thread-4" RUNNING
                                                     {1004739B53}>:
  Execution of a form compiled with errors.
Form:
  ("RX" "RZ")
Compile-time error:
  illegal function call

1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047B9123}>)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047B9123}>)
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:COMPILED-PROGRAM-ERROR {10047B7433}> #<unused argument> :QUIT T)01: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047B9123}>)
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:COMPILED-PROGRAM-ERROR {10047B7433}> #<unused argument> :QUIT T)0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:COMPILED-PROGRAM-ERROR {10047A3053}> #<unused argument> :QUIT T)0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:COMPILED-PROGRAM-ERROR {10047AA6B3}> #<unused argument> :QUIT T)2: (INVOKE-DEBUGGER #<SB-INT:COMPILED-PROGRAM-ERROR {10047B9123}>)
RUNNING {1004739B53}>
3: (ERROR SB-INT:COMPILED-PROGRAM-ERROR :MESSAGE "illegal function call" :SOURCE "(\"RX\" \"RZ\")")
2: (INVOKE-DEBUGGER #<SB-INT:COMPILED-PROGRAM-ERROR {10047AA6B3}>)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047A3053}>)1: (SB-DEBUG::RUN-HOOK *INVOKE-1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047A3053}>)1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047B7433}>)
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:COMPILED-PROGRAM-ERROR {10047CF0F3}> #<unused argument> :QUIT T)
DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047B7433}>)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047A3053}>)1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047B7433}>)
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:COMPILED-PROGRAM-ERROR {10047CF0F3}> #<unused argument> :QUIT T)
3: (ERROR SB-INT:COMPILED-PROGRAM-ERROR :MESSAGE "illegal function call" :SOURCE "(\"RX\" \"RZ\")")
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047A3053}>)1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:COMPILED-PROGRAM-ERROR {10047B7433}>)
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:COMPILED-PROGRAM-ERROR {10047CF0F3}> #<unused argument> :QUIT T)
3: (ERROR SB-INT:COMPILED-PROGRAM-ERROR :MESSAGE "illegal function call" :SOURCE "(\"RX\" \"RZ\")")
2: (INVOKE-DEBUGGER #<SB-INT:COMPILED-PROGRAM-ERROR {10047B7433}>)
```

File: app/src/entry-point.lisp [hunk 1]
```diff
@@ -331,7 +331,7 @@ end of the name."
                   :collect `(,name ,value))
           (bordeaux-threads:*default-special-bindings*
             (list* ,@(loop :for (name value) :in let-defs
-                           :collect `(cons ',name ,name))
+                           :collect ``(,',name . ',,name))
```
Comment (ecpeterson) [inline app/src/entry-point.lisp]: I think I understand what's going on here, but this spooks me a little

Comment (appleby) [reply app/src/entry-point.lisp]: :-)

Indeed, expert opinion is welcome. I don't break out the double-backquote that often!

The crux of it is that we want two levels of unquote. That is, we want to evaluate `name` at macro-expansion time to get the symbol inside it, and we want to evaluate *that* symbol in the expanded form (at compile/load/whatever time), but to ensure that it's value is quoted.

In other words, we want to expand into a backquoted form like 

```lisp
`(*foo* . ',*foo*)
```

but due to idiosyncrasies of multiple-unquote, the best I could muster was:

```lisp
`(,'*foo* . ',*foo*)
```

Comment (rigettizach) [reply app/src/entry-point.lisp]: I think this suffices and is a little easier to follow (for me):

    :collect `(cons ',name (list 'quote ,name))

Comment (appleby) [reply app/src/entry-point.lisp]: > I think this suffices and is a little easier to follow (for me):
> 
> ```
> :collect `(cons ',name (list 'quote ,name))
> ```

See, this is why ecpeterson summoned you :).

I agree, your version is easier to follow. Tests running now. Will push an update soon.
